configurationDone signature incorrectly says "Void"

vscode-js-debug sends the following response to configurationDone:
```json
{
  "jsonrpc": "2.0",
  "id": 4,
  "method": "configurationDone",
  "params": {}
}
```
This is legal according to the specification.
However, this cause LSP4J to fail at parsing the message, because LSP4J mandates a `Void`. The exception is
```
Caused by: java.lang.ClassCastException: class com.google.gson.JsonObject cannot be cast to class java.lang.Void (com.google.gson.JsonObject is in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @694b1ddb; java.lang.Void is in module java.base of loader 'bootstrap')
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleResponse(RemoteEndpoint.java:212)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:193)
	at org.eclipse.lsp4j.jsonrpc.validation.ReflectiveMessageValidator.consume(ReflectiveMessageValidator.java:68)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:194)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:577)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1589)
```

--- a/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/DebugIntegrationTest.java
+++ b/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/DebugIntegrationTest.java
@@ -56,6 +56,12 @@ public class DebugIntegrationTest {
 		CompletableFuture<MyParam> askServer(MyParam param);
 	}
 
+
+	public static interface MyVoidServer {
+		@JsonRequest
+		CompletableFuture<Void> askServer(MyParam param);
+	}
+
 	public static interface MyClient {
 		@JsonRequest
 		CompletableFuture<MyParam> askClient(MyParam param);
@@ -97,6 +103,45 @@ public class DebugIntegrationTest {
 
 		Assert.assertEquals("FOO", fooFuture.get(TIMEOUT, TimeUnit.MILLISECONDS).value);
 		Assert.assertEquals("BAR", barFuture.get(TIMEOUT, TimeUnit.MILLISECONDS).value);
+	}
+
+	@Test
+	public void testVoidResponse() throws Exception {
+		// create client side
+		PipedInputStream in = new PipedInputStream();
+		PipedOutputStream out = new PipedOutputStream();
+		PipedInputStream in2 = new PipedInputStream();
+		PipedOutputStream out2 = new PipedOutputStream();
+
+		in.connect(out2);
+		out.connect(in2);
+
+		MyClient client = new MyClient() {
+			@Override
+			public CompletableFuture<MyParam> askClient(MyParam param) {
+				throw new UnsupportedOperationException("Unused by this test");
+			}
+		};
+		Launcher<MyVoidServer> clientSideLauncher = DebugLauncher.createLauncher(client, MyVoidServer.class, in, out);
+
+		// create server side
+		MyServer server = new MyServer() {
+			@Override
+			public CompletableFuture<MyParam> askServer(MyParam param) {
+				return CompletableFuture.completedFuture(param);
+			}
+		};
+		Launcher<MyClient> serverSideLauncher = DebugLauncher.createLauncher(server, MyClient.class, in2, out2);
+
+		clientSideLauncher.startListening();
+		serverSideLauncher.startListening();
+
+		// We call a method that is declared as returning Void, but the other end returns a non-null value
+		// make sure that the json parsing discards that result
+		CompletableFuture<Void> fooFuture = clientSideLauncher.getRemoteProxy().askServer(new MyParam("FOO"));
+
+		Void void1 = fooFuture.get(TIMEOUT, TimeUnit.MILLISECONDS);
+		Assert.assertNull(void1);
 	}
 
 	@Test

--- a/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/json/DebugMessageJsonHandlerTest.java
+++ b/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/json/DebugMessageJsonHandlerTest.java
@@ -850,7 +850,7 @@ public class DebugMessageJsonHandlerTest {
 		}
 	}
 
-	private Object voidResponse(String body) {
+	private void voidResponse(String body) {
 		Map<String, JsonRpcMethod> supportedMethods = new LinkedHashMap<>();
 		supportedMethods.put("foo", JsonRpcMethod.request("foo", new TypeToken<Void>() {
 		}.getType(), new TypeToken<Void>() {
@@ -870,56 +870,42 @@ public class DebugMessageJsonHandlerTest {
 				+ "\"success\":true"//
 				+ bodyField //
 				+ "}\n");
-		return ((ResponseMessage) message).getResult();
+		Object result = ((ResponseMessage) message).getResult();
+		Assert.assertNull(result);
 	}
 
 	@Test
 	public void testVoidResponse_noBody() {
-		Assert.assertNull(voidResponse(null));
+		voidResponse(null);
 	}
 
 	@Test
 	public void testVoidResponse_null() {
-		Assert.assertNull(voidResponse("null"));
-
+		voidResponse("null");
 	}
 
 	@Test
 	public void testVoidResponse_primitive() {
-		Object result = voidResponse("true");
-		JsonPrimitive jsonPrimitive = (JsonPrimitive) result;
-		Assert.assertTrue(jsonPrimitive.getAsBoolean());
+		voidResponse("true");
 	}
 
 	@Test
 	public void testVoidResponse_emptyArray() {
-		Object result = voidResponse("[]");
-		JsonArray jsonArray = (JsonArray) result;
-		Assert.assertTrue(jsonArray.isEmpty());
+		voidResponse("[]");
 	}
 
 	@Test
 	public void testVoidResponse_array() {
-		Object result = voidResponse("[1,2,3]");
-		JsonArray jsonArray = (JsonArray) result;
-		Assert.assertEquals(1, jsonArray.get(0).getAsInt());
-		Assert.assertEquals(2, jsonArray.get(1).getAsInt());
-		Assert.assertEquals(3, jsonArray.get(2).getAsInt());
+		voidResponse("[1,2,3]");
 	}
 
 	@Test
 	public void testVoidResponse_emptyObject() {
-		Object result = voidResponse("{}");
-		Assert.assertTrue(result instanceof JsonObject);
-		JsonObject jsonObject = (JsonObject) result;
-		Assert.assertTrue(jsonObject.entrySet().isEmpty());
+		voidResponse("{}");
 	}
 
 	@Test
 	public void testVoidResponse_object() {
-		Object result = voidResponse("{\"allThreadsContinued\":false}");
-		Assert.assertTrue(result instanceof JsonObject);
-		JsonObject jsonObject = (JsonObject) result;
-		Assert.assertFalse(jsonObject.getAsJsonPrimitive("allThreadsContinued").getAsBoolean());
+		voidResponse("{\"allThreadsContinued\":false}");
 	}
 }

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapter.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/adapters/MessageTypeAdapter.java
@@ -350,7 +350,7 @@ public class MessageTypeAdapter extends TypeAdapter<Message> {
 			return null;
 		}
 		if (isNullOrVoidType(type)) {
-			return element;
+			return null;
 		}
 		Object value = gson.fromJson(element, type);
 		if (isNull(value)) {

--- a/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/IntegrationTest.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/IntegrationTest.java
@@ -31,6 +31,7 @@ import org.eclipse.lsp4j.jsonrpc.CompletableFutures;
 import org.eclipse.lsp4j.jsonrpc.Launcher;
 import org.eclipse.lsp4j.jsonrpc.RemoteEndpoint;
 import org.eclipse.lsp4j.jsonrpc.ResponseErrorException;
+
 import org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
 import org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint;
@@ -87,6 +88,11 @@ public class IntegrationTest {
 	public static interface MyServer {
 		@JsonRequest
 		CompletableFuture<MyParam> askServer(MyParam param);
+	}
+
+	public static interface MyVoidServer {
+		@JsonRequest
+		CompletableFuture<Void> askServer(MyParam param);
 	}
 	
 	public static class MyServerImpl implements MyServer {
@@ -222,6 +228,45 @@ public class IntegrationTest {
 		Assert.assertEquals("Content-Length: 50" + CRLF + CRLF
 				+ "{\"jsonrpc\":\"2.0\",\"id\":42,\"result\":{\"value\":\"foo\"}}",
 				out.toString());
+	}
+
+	@Test
+	public void testVoidResponse() throws Exception {
+		// create client side
+		PipedInputStream in = new PipedInputStream();
+		PipedOutputStream out = new PipedOutputStream();
+		PipedInputStream in2 = new PipedInputStream();
+		PipedOutputStream out2 = new PipedOutputStream();
+
+		in.connect(out2);
+		out.connect(in2);
+
+		MyClient client = new MyClient() {
+			@Override
+			public CompletableFuture<MyParam> askClient(MyParam param) {
+				throw new UnsupportedOperationException("Unused by this test");
+			}
+		};
+		Launcher<MyVoidServer> clientSideLauncher = Launcher.createLauncher(client, MyVoidServer.class, in, out);
+
+		// create server side
+		MyServer server = new MyServer() {
+			@Override
+			public CompletableFuture<MyParam> askServer(MyParam param) {
+				return CompletableFuture.completedFuture(param);
+			}
+		};
+		Launcher<MyClient> serverSideLauncher = Launcher.createLauncher(server, MyClient.class, in2, out2);
+
+		clientSideLauncher.startListening();
+		serverSideLauncher.startListening();
+
+		// We call a method that is declared as returning Void, but the other end returns a non-null value
+		// make sure that the json parsing discards that result
+		CompletableFuture<Void> fooFuture = clientSideLauncher.getRemoteProxy().askServer(new MyParam("FOO"));
+
+		Void void1 = fooFuture.get(TIMEOUT, TimeUnit.MILLISECONDS);
+		Assert.assertNull(void1);
 	}
 
 

